sim/data_providers.py

```diff
@@ -1,84 +1,70 @@
-#pylint: too-many-arguments, too-many-locals
+# pylint: disable=too-many-arguments, too-many-locals
 
 import os
 from typing import (
     List,
-    Dict,
     Optional,
     Any,
+    Union,
 )
 import qcodes as qc
 import xarray as xr
 from scipy import interpolate
 from sim.qcodes_utils import QcodesDbConfig
+from sim.mock_pin import IMockPin
+from sim.data_provider import IDataProvider
+from sim.mock_device_registry import MockDeviceRegistry
 
 
-class Pin:
-    def __init__(
-        self,
-        name: str,
-    ) -> None:
-        self._name = name
-        self._value: float = 0.0
+class DataProvider(IDataProvider):
+    """Base class for data providers"""
 
-    def __repr__(self) -> str:
-        return self._name
-
-    def __str__(self) -> str:
-        return self._name
+    def __init__(self, settable: bool):
+        self._settable = settable
 
     @property
-    def name(self) -> str:
-        """Name of the pin"""
-
-        return self._name
-
-    @property
-    def value(self) -> float:
-        """Gets the current value of the pin"""
-
-        return self._value
-
-    def get_value(self) -> float:
-        """Gets the current value on the input pin.  Compatible with qcodes
-        Parameter get_cmd argument.
+    def settable(self) -> bool:
+        """Indicates whether this data provider allows its value to
+        be set by calling set_value
         """
-
-        return self._value
+        return self._settable
 
 
-class StaticDataProvider:
+class StaticDataProvider(DataProvider):
     """Data provider that returns a constant value for all inputs."""
 
     def __init__(self, value: float) -> None:
-        """Initializes a static data provider."""
-
+        super().__init__(settable=True)
         self._value = value
-        self._xarray = xr.DataArray("0", dims="x", coords={"x":[1]})
 
     def __call__(self, *args) -> float:
         return self._value
 
+    def set_value(self, value: Any) -> None:
+        """Set the static value of this data provider"""
+        self._value = value
+
     @property
     def value(self) -> float:
-        """ """
+        """The current value of this data provider"""
 
         return self._value
 
+    @property
     def raw_data(self) -> xr.DataArray:
         """Returns the raw data backing this provider as an
         xarray.DataArray
         """
 
-        return self._xarray
+        return xr.DataArray("0", dims="x", coords={"x": [1]})
 
 
-class QcodesDataProvider:
+class QcodesDataProvider(DataProvider):
     """Data provider that sources it's data from a 1D or 2D QCoDeS dataset."""
 
     def __init__(
         self,
-        input_providers: List[Pin],
+        input_providers: List[Union[str, IMockPin]],
```

Comment:
In general I would favour the Abstract collections type Sequence over List as in input argument to a function since:

* It allows the user to pass both tuples, lists and other types that implement indexing
* It makes it clear to the type system and the reader that the method does not mutate it input.

Reply:
Good suggestion.